apps/backend/src/authn/github.strategy.ts

```diff
@@ -0,0 +1,84 @@
+import {Injectable, UnauthorizedException} from '@nestjs/common';
+import {PassportStrategy} from '@nestjs/passport';
+import axios from 'axios';
+import {Strategy} from 'passport-github';
+import {ConfigService} from '../config/config.service';
+import {User} from '../users/user.model';
+import {AuthnService} from './authn.service';
+
+interface GithubProfile {
+  name: string;
+}
+
+interface GithubEmail {
+  email: string;
+  verified: boolean;
+}
+
+@Injectable()
+export class GithubStrategy extends PassportStrategy(Strategy, 'github') {
+  constructor(
+    private readonly authnService: AuthnService,
+    private readonly configService: ConfigService
+  ) {
+    super({
+      clientID: configService.get('GITHUB_CLIENTID') || 'disabled',
+      clientSecret: configService.get('GITHUB_CLIENTSECRET') || 'disabled',
+      scope: 'user:email',
+      passReqToCallback: true
+    });
+  }
+
+  async validate(
+    req: Record<string, unknown>,
+    accessToken: string
+  ): Promise<User> {
+    // Get user's linked emails from Github
+    const githubEmails = await axios
+      .get(
+        this.configService.get('CYPRESS_TESTING') === 'true'
+          ? 'http://127.0.0.1:3001/emails'
+          : 'https://api.github.com/user/emails',
+        {
```

Comment:
For more flexibility here, should the value come from the ENV? Basically if `CYPRESS_TESTING_CALLBACK` is set then use the URL stored in `CYPRESS_TESTING_CALLBACK`, else use `https://api.github.com/user/emails`

Reply:
To allow for enterprise GitHub instances, I'm adding GITHUB_ENTERPRISE.. INSTANCE_URL, TOKEN_URL, PROFILE_URL, and EMAIL_URL which will default to public github, and will be set to the mock server during testing.